### DISCUSSION TITLE
Auditing EF8 Core Complex Properties

### DISCUSSION THREAD

**Discussion by NickNightingale** [thepirat000/Audit.NET#697] [Q&A] [answered]
I am a new user of Audit.EntityFramework.Core. I am auditing quite a simple database using AuditTypeNameMapper and mostly everything is working really well. I have a few complex properties (e.g. a struct with X and Y properties)  and while I can see these in the EntityFrameworkEvent separated out into their generated column names (e.g. Element_X and Element_Y), they do not get copied automatically into the matching audit record. Is this a limitation of the column matching? If so is there a better way to handle this type of property?

**Comment by thepirat000** [comment]
I assume you're referring to an audit log entity where the column names of complex properties correspond directly to the audit entity properties.

```c#
[ComplexType]
class Element 
{ 
  public int X { get; set; } 
  public int Y { get; set; } 
}

// Entity to audit
class Foo
{
    [Key]
    public int Id { get; set; }
    public string Name { get; set; }
    public Element Element { get; set; }
}

// Audit entity with 'matching' properties
class FooAuditLog
{
    [Key]
    public int AuditId { get; set; }

    public string Name { get; set; }

    [Required]
    public Element Element { get; set; } // <--- Is this what you're asking about?
}
```

If that's what you're referring to, then yes, currently the library will not automatically match complex properties.

[This line](https://github.com/thepirat000/Audit.NET/blob/e382cf6fffe3e54bd728e4b750f9cf3173ab821f/src/Audit.EntityFramework/Providers/EntityFrameworkDataProvider.cs#L314) only returns scalar properties, excluding complex ones.

I'll examine this further, as I believe the issue can be fixed by adding the mapping of the complex properties from the entity to the audit entity.

**Comment by thepirat000** [comment, marked as answer]
This was fixed in version 27.0.3. Kindly update your references to the latest version and re-test. If the issue persists, feel free to leave a comment below

**Reply by NickNightingale** [reply]
Thank you for the really fast, detailed response and apologies for not including a code sample. This was exactly my question and after updating to 27.0.3 it works perfectly. Thank you for the excellent library!
